pandas/core/algorithms.py

```diff
@@ -728,25 +733,31 @@ def factorize(
 
     if not isinstance(values.dtype, np.dtype):
         # i.e. ExtensionDtype
+        assert dropna or sort
         codes, uniques = values.factorize(na_sentinel=na_sentinel)
```

Comment:
Currently not hit by any tests, but I think would be if e.g Int64 was used in the grouper. This will need fixing.